### PR TITLE
rgw: add support external iam authorization

### PR DESCRIPTION
**Pull request by clwluvw**
Implemented `RGWExternIAMAuthorize` to issue a request with the `rgw::IAM::Environment`, `rgw::ARN`, `op`, and `rgw_user` info to an external service to check for the `rgw::IAM::Effect` result.

Fixes: https://tracker.ceph.com/issues/62775

### CODE REVIEW

**Comment by mattbenjamin** [comment]
if we were going to permit this, shouldn't it be implemented as an auth strategy?  @clwluvw does that make sense?

**Comment by clwluvw** [comment]
> if we were going to permit this, shouldn't it be implemented as an auth strategy? @clwluvw does that make sense?

I agree with you (I wasn't really happy with the Singleton class I put for this here). I guess the best is to make it an auth strategy so that later it can be extended to authenticate as well.
Right now I guess we only carry `rgw::auth::Identity` with `req_state` on RGWOP funcs. But the Appliers implemented so far don't have any sort of authz function. For example `RoleApplier` only adds policies to the `req_state` and then again there is a function to authz on `iam_user_policies` if there are any.
So if I'm not mistaken maybe auth strategies should have a sort of "authz" func that can be called in the middle of RGWOP flows (like with the same signature that `eval` has)? 🤔 
Also need to consider how we can mix other sort of authentication/authorization with it (I mean we shouldn't lose sts or bucket policy by enabling this for example)...

Will take a deeper look.

**Comment by clwluvw** [comment]
@mattbenjamin I took a deeper look. What I see in `rgw::auth::Strategy` happening there is doing the authentication and returning a pair of `IdentityApplier` and `Completer` in the `apply` function.
The current authorization flows are "IAM Policies" (aka. bucket policies - loaded always by `RGWHandler::do_init_permissions`), "IAM User Policies" and "Session Policies" (loaded by `rgw::auth::RoleApplier::modify_request_state` into `req_state`) and "ACLs" (which is based on the `rgw::auth::Identity` to check for the owner and...).

The fact is the "IAM authorization" is happening in the middle of each `RGWOP` by checking if for example there is any session policy to `eval` or not. (So what I'm trying to point out is there is nothing `rgw::auth::Strategy` or `rgw::auth::IdentityApplier`  related calling there).
So based on the available authorization methods I mentioned above - ACL and Bucket policies are always there regardless of the `auth_strategy.authenticate()` result and only "IAM User Policies" and "Session Policies" are _loaded_ if the auth strategy was STS.

For this particular feature ("External IAM Authorization") - We need to have it in addition to all authorization mechanisms we have IMHO. That means If the authentication happened with STS or Local or Remote - the Authz still should be able to work with External IAM (if enabled) and other internals (like sts policies). (Like how it is implemented in this PR right now)

To have something in addition to all available mechanisms regardless of the authentication engine - We probably should follow the same arch of modifying `req_state` and put something there (like what STS does) and call it if it was not "NULL" - Which I guess it will be fallback to the current implementation I did but just an object in req_state rather than a Singleton class somewhere else.

Even if we add authentication enabled for this "External IAM" - Again we would need to have this "External IAM Authorization" enabled always because the authentication can fall back on other engines/strategies and then we want to have it enabled with all.

So a sum up would be probably to either add a class object every time to `req_state` in `rgw::auth::Strategy::apply` or keep it like what is implemented in this PR...

Please let me know if I'm mistaken anywhere and let me know what you think.

**Comment by cbodley** [comment]
hey @clwluvw, can you please share your motivations for hooking iam auth this way? is this related to earlier discussions about the open policy agent? the description in https://tracker.ceph.com/issues/62775 is short on details

**Comment by clwluvw** [comment]
@cbodley I put more info to the ticket. It's somehow related to the concept that OPA implemented but not like that because with OPA you cannot Allow a request that is denied by ACLs and you don't have all the info (IAM policy doc info) in the request.

**Comment by cbodley** [comment]
> Current External Auth implementations (keystone, ldap) don't offer any authorization (IAM policy doc-like).

@clwluvw what are the gaps with IAM policy and keystone auth? can a bucket policy not specify a `Principal` that refers to a keystone user? can you not add user policies to the 'shadow user' associated with a keystone user?

cc @adamemerson @pritha-srivastava

**Comment by cbodley** [comment]
> Having a free protocol to send the IAM info to an external service and do the authorization for the request was always a missing feature in Ceph (RGW).

@clwluvw isn't the interpretation and application of IAM policy supposed to be deterministic? why would we need an external service for this, if we expect rgw to make these policy decisions correctly?

**Comment by clwluvw** [comment]
@cbodley It's not about the principal in bucket policies. It's about having the user policy in addition to the bucket policy from an external IAM service which might have different language in policy docs but understands AWS S3 policy tags.
The difference between bucket policy and user policy is more that one can [have a single policy for many buckets or many services in the cloud](https://aws.amazon.com/blogs/security/iam-policies-and-bucket-policies-and-acls-oh-my-controlling-access-to-s3-resources/) or have the [understanding of the region or...](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference-arns.html) (which are most "cloud/company" specific).
I mean just exactly how bucket policy and user policy in AWS or GCP or... are working.
Basically in this video (https://youtu.be/YMj33ToS8cI?si=YWrjMXFp8ierBxIS), it was perfectly explained (it also has S3 as an example in it) how this "external IAM" (which is not S3 specific) is useful and can let users make more policies based on other products in the cloud.

In a nutshell, the fact is to let the S3 product ask a central (external) service to check whether this request is allowed/denied/passed based on its policies in addition to its bucket policies. the thing you say, to me, is like asking that external service to push policies as a bucket policy to Ceph which with all the notes above I guess it's not possible and makes the system a lot complex in terms of consistency and translation of services and...

**Comment by mattbenjamin** [comment]
So this is the OPA product insertion once again.

**Comment by mattbenjamin** [comment]
I don't find this persuasive, as given.  A more plausible model, to me, is the one exemplified in STS.  When a user authenticates with its SSO, it may acquire a flexible set of authorization vectors that can interact with user and especially bucket policy.  Taking advantage of these does not require the S3 service to make live callouts to an external system to make authorization decisions, it can make them discretely and on its own.  The supervising authority, whatever that is, controls both policy and the propagation of those vectors--and both are part of a web of security infrastructure/trust whose rules and security model are well known.  A system that acts like a remote oracle is both a security risk to RGW's operation, communicating with one risks shoveling privileged information to a system RGW knows nothing about.

**Comment by clwluvw** [comment]
> A system that acts like a remote oracle is not only a security risk to RGW's operation, communicating with one risks shoveling privileged information to a system RGW knows nothing about.

@mattbenjamin I guess what you see as a kind of bug here is a feature now in the globe :D
I mean this concept of having central IAM authorization in cloud providers (not only AWS, you see in GCP, Azure, ...) is acting the same that products (S3, DBaaS, EC2, ...) ask this service whether should I allow/deny the request (for example here is HTTP request in other products is others), and then they blindly do it because they trust the system by it means (as what that service was designed for).

**Comment by clwluvw** [comment]
I can also refer to this video (https://youtu.be/tPr1AgGkvc4?si=xqEQWYol52mXBnsV) as how they are dealing with such a thing at scale...

**Comment by pritha-srivastava** [comment]
I agree with @mattbenjamin and @cbodley , that rgw already provides an IAM and a STS service (in addition to s3), then why do we need an external iam authorization mechanism here. There might be functionalities missing/ untested - like specifying a keystone/ ldap user as a Principal in a bucket policy - we should be looking into them instead. Also if OPA already provides external authorization and has some things missing, probably that can be enhanced. I have a question here though - allowing requests by externally authorizing them even when an existing ACL denies it - is this even valid, because in AWS the request is explicitly denied in case any of the policies deny it.

**Comment by clwluvw** [comment]
> I agree with @mattbenjamin and @cbodley , that rgw already provides an IAM and a STS service (in addition to s3), then why do we need an external iam authorization mechanism here. There might be functionalities missing/ untested - like specifying a keystone/ ldap user as a Principal in a bucket policy - we should be looking into them instead.

@pritha-srivastava I'm not sure if I understand the conclusion here correctly... You're comparing two different features here. External IAM is something "company-specific" - a real example for better understanding is the [AWS User Policy](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html), [GCP Policy](https://cloud.google.com/iam/docs/reference/rest/v1/Policy) - and bucket policy is something s3 specific which applies on bucket level.
The STS and IAM you're referring to here also is something inside the S3 cluster (product) - which is not what I'm trying to solve here as a feature. (i.e. these are two different IAM strategies)

> I have a question here though - allowing requests by externally authorizing them even when an existing ACL denies it - is this even valid, because in AWS the request is explicitly denied in case any of the policies deny it.

You are saying policies, not ACLs. With policies, you can share access to your resources with others (i.e. you'll bypass the ACL check here). This is the same implementation we have for bucket policy and IAM policies for Roles in Ceph currently and the External IAM I proposed here is also following the same which is correct.

**Comment by clwluvw** [comment]
Unstale please

**Comment by clwluvw** [comment]
Unstale please

**Comment by clwluvw** [comment]
unstale please.

**Comment by clwluvw** [comment]
unstale please.

**Comment by clwluvw** [comment]
unstale please
